### PR TITLE
[DataGrid] Spotting perf issues

### PR DESCRIPTION
**Pull request by dtassone**
This PR aims to
- fix some perf bottleneck and refactor the selection.
- Also fix an issue with the sorting

### CODE REVIEW

File: packages/grid/_modules_/grid/hooks/root/useApiMethod.ts [hunk 1]
```diff
@@ -21,7 +21,13 @@ export function useApiMethod(apiRef: ApiRef, apiMethods: Partial<GridApi>, apiNa
...
         logger.debug(`Adding ${apiName}.${methodName} to apiRef`);
         apiRef.current[methodName] = (...args) => {
-          return apiMethodsRef.current[methodName](...args);
+          const start = new Date().getTime();
+          const result = apiMethodsRef.current[methodName](...args);
+          const end = new Date().getTime();
+          if(end - start > 10) {
+            console.log(`${methodName} took ${end - start}`);
+          }
+          return result;
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/hooks/root/useApiMethod.ts:30]: Did you intend to merge the performance diff or was it to debug?

**Review by oliviertassinari — CHANGES_REQUESTED** [review]
The test surface a regression, sorting is no longer working

File: packages/grid/_modules_/grid/hooks/root/useApiMethod.ts (hunk 1 above)
Comment (dtassone) [reply packages/grid/_modules_/grid/hooks/root/useApiMethod.ts:30]: No it's just for debugging

**Comment by DanailH** [comment]
> The tests surface a regression, sorting is no longer working on the first click. You can reproduce on https://deploy-preview-501--material-ui-x.netlify.app/components/data-grid/rows/#basic-sorting. It's actually great that we are doing this refactorization, each bug is an opportunity to add a new test case.

I had the same problem after I removed the `React.memo` from the `column-header-item`

File: packages/grid/_modules_/grid/models/api/sortApi.ts [hunk 2]
```diff
@@ -9,6 +9,10 @@ export interface SortApi {
    * Get the sort model currently applied to the grid.
    */
   getSortModel: () => SortModel;
+  /**
+   * Apply the current sorting model to the rows;
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/models/api/sortApi.ts:13]:
```suggestion
   * Apply the current sorting model to the rows.
```

File: packages/grid/_modules_/grid/hooks/root/useContainerProps.ts [hunk 3]
```diff
@@ -41,7 +41,8 @@ export const useContainerProps = (windowRef: React.RefObject<HTMLDivElement>, ap
   }, [options.pagination, paginationState.page, paginationState.pageSize, totalRowsCount]);
 
   const getContainerProps = React.useCallback((): ContainerProps | null => {
-    if (!windowRef || !windowRef.current || columnsTotalWidth === 0) {
+    // eslint-disable-next-line no-restricted-globals
+    if (!windowRef || !windowRef.current || columnsTotalWidth === 0 || isNaN(columnsTotalWidth)) {
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/hooks/root/useContainerProps.ts:45]:
```suggestion
    if (!windowRef || !windowRef.current || columnsTotalWidth === 0 || Number.isNaN(columnsTotalWidth)) {
```

File: packages/grid/_modules_/grid/models/params/rowSelectedParams.ts [hunk 4]
```diff
@@ -8,10 +8,7 @@ export interface RowSelectedParams {
    * The row data of the row that triggers the event.
    */
   data: RowData;
-  /**
-   * The row index of the row that triggers the event.
-   */
-  rowIndex: number;
+
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/models/params/rowSelectedParams.ts:11]:
```suggestion
```

File: packages/grid/_modules_/grid/models/colDef/checkboxSelection.tsx [hunk 5]
```diff
@@ -1,8 +1,7 @@
 import * as React from 'react';
-import { ColDef } from './colDef';
 import { CellCheckboxRenderer, HeaderCheckbox } from '../../components/checkbox-renderer';
 
-export const checkboxSelectionColDef: ColDef = {
+export const checkboxSelectionColDef: any = {
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/models/colDef/checkboxSelection.tsx:4]: any?

**Review by oliviertassinari — APPROVED** [review]
> fix an issue with the sorting

Is the bug something we can/should add a test case for?

File: packages/grid/_modules_/grid/models/colDef/checkboxSelection.tsx (hunk 5 above)
Comment (dtassone) [reply packages/grid/_modules_/grid/models/colDef/checkboxSelection.tsx:4]: Yes I had a cyclic dep issue, I will try again

**Comment by dtassone** [comment]
Any idea @oliviertassinari why karma is not happy? Or is it just personal 😛

**Comment by oliviertassinari** [comment]
> why karma is not happy? 

The tests seem to pass, what was the error?

**Comment by oliviertassinari** [comment]
Oh yeah right, there is an infinite loop with the data grid something is wrong in the logic.

**Comment by oliviertassinari** [comment]
I could pinpoint the test that fails. it's `prop: checkboxSelection > should check and uncheck when double clicking the row`.

**Comment by oliviertassinari** [comment]
There are two problems:

1. The click event is no longer act-synchronously handled, hence the assertion fails. This is probably because we now relies on a RAF that [React's act](https://reactjs.org/docs/testing-recipes.html#act) can't handle. So we miss a `await raf()`.
```diff
diff --git a/packages/grid/x-grid/src/XGrid.test.tsx b/packages/grid/x-grid/src/XGrid.test.tsx
index b5da9b88..a7e9c0f4 100644
--- a/packages/grid/x-grid/src/XGrid.test.tsx
+++ b/packages/grid/x-grid/src/XGrid.test.tsx
@@ -189,7 +189,7 @@ describe('<XGrid />', () => {
   });

   describe('prop: checkboxSelection', () => {
-    it('should check and uncheck when double clicking the row', async () => {
+    it.only('should check and uncheck when double clicking the row', async () => {
       render(
         <div style={{ width: 300, height: 300 }}>
           <XGrid
@@ -213,10 +213,12 @@ describe('<XGrid />', () => {
       expect(checkbox).to.have.property('checked', false);

       fireEvent.click(screen.getByRole('cell', { name: 'Nike' }));
+      await raf();
       expect(row).to.have.class('Mui-selected');
       expect(checkbox).to.have.property('checked', true);

       fireEvent.click(screen.getByRole('cell', { name: 'Nike' }));
+      await raf();
       expect(row).to.not.have.class('Mui-selected');
       expect(checkbox).to.have.property('checked', false);
     });
```
2. The second is really weird and worrying. It's the first time I see it, I struggle to reproduce it in a clean sheet. The `expect(row).to.have.class('Mui-selected');` assertion crashes. On the other hand, this correctly fail. @eps1lon have you ever seen something like this? :
```diff
diff --git a/packages/grid/x-grid/src/XGrid.test.tsx b/packages/grid/x-grid/src/XGrid.test.tsx
index b5da9b88..5c1f4a08 100644
--- a/packages/grid/x-grid/src/XGrid.test.tsx
+++ b/packages/grid/x-grid/src/XGrid.test.tsx
@@ -189,7 +189,7 @@ describe('<XGrid />', () => {
   });

   describe('prop: checkboxSelection', () => {
-    it('should check and uncheck when double clicking the row', async () => {
+    it.only('should check and uncheck when double clicking the row', async () => {
       render(
         <div style={{ width: 300, height: 300 }}>
           <XGrid
@@ -213,11 +213,11 @@ describe('<XGrid />', () => {
       expect(checkbox).to.have.property('checked', false);

       fireEvent.click(screen.getByRole('cell', { name: 'Nike' }));
-      expect(row).to.have.class('Mui-selected');
+      expect(row!.classList.contains('Mui-selected')).to.equal(true);
       expect(checkbox).to.have.property('checked', true);

       fireEvent.click(screen.getByRole('cell', { name: 'Nike' }));
-      expect(row).to.not.have.class('Mui-selected');
+      expect(row!.classList.contains('Mui-selected')).to.equal(false);
       expect(checkbox).to.have.property('checked', false);
     });
   });
```

**Comment by oliviertassinari** [comment]
`.have.class`'s implementation [is straightforward](https://github.com/nathanboktae/chai-dom/blob/75c0e73b1dcce2ee39f8e24ace33bf9c6292d858/chai-dom.js#L77), so maybe it's the serialization of the DOM node to a string that crashes.

**Comment by oliviertassinari** [comment]
We can investigate more if it reproduces more often.

**Comment by eps1lon** [comment]
> The click event is no longer act-synchronously handled, hence the assertion fails. This is because we now relies on a RAF that

Why would it rely on requestAnimationFrame?

**Comment by oliviertassinari** [comment]
> Why would it rely on requestAnimationFrame?

@eps1lon I doubt that the issue with the test is related to the RAF. I suspect that the DOM structure of the row has something that chai-dom doesn't like, but it's only a hypothesis. For why a RAF we can chat more about it in #507.

**Pull request merged**